server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java

```diff
@@ -210,12 +216,151 @@ public void testPaginationRequiresVerboseListing() throws Exception {
+        assertThat(clusterAdmin().prepareGetSnapshots("other*", "-o*").setSnapshots(matchAllPattern()).get().getSnapshots(), empty());
+    }
+
+    public void testNamesStartingInDash() {
+        final String repoName1 = "test-repo";
+        final String weirdRepo1 = "-weird-repo-1";
+        final String weirdRepo2 = "-weird-repo-2";
+        createRepository(repoName1, "fs");
+        createRepository(weirdRepo1, "fs");
+        createRepository(weirdRepo2, "fs");
+
+        final String snapshotName = "test-snapshot";
+        final String weirdSnapshot1 = "-weird-snapshot-1";
+        final String weirdSnapshot2 = "-weird-snapshot-2";
+
+        final SnapshotInfo snapshotInRepo1 = createFullSnapshot(repoName1, snapshotName);
+        final SnapshotInfo weirdSnapshot1InRepo1 = createFullSnapshot(repoName1, weirdSnapshot1);
+        createFullSnapshot(repoName1, weirdSnapshot2);
+
+        final SnapshotInfo snapshotInWeird1 = createFullSnapshot(weirdRepo1, snapshotName);
+        final SnapshotInfo weirdSnapshot1InWeird1 = createFullSnapshot(weirdRepo1, weirdSnapshot1);
+        createFullSnapshot(weirdRepo1, weirdSnapshot2);
+
+        final SnapshotInfo snapshotInWeird2 = createFullSnapshot(weirdRepo2, snapshotName);
+        final SnapshotInfo weirdSnapshot1InWeird2 = createFullSnapshot(weirdRepo2, weirdSnapshot1);
+        createFullSnapshot(weirdRepo2, weirdSnapshot2);
+
+        final List<SnapshotInfo> allSnapshots = clusterAdmin().prepareGetSnapshots(matchAllPattern())
+            .setSort(GetSnapshotsRequest.SortBy.REPOSITORY)
+            .get()
+            .getSnapshots();
+        assertThat(allSnapshots, hasSize(9));
+
+        final List<SnapshotInfo> allSnapshotsByAll = getAllByPatterns(matchAllPattern(), matchAllPattern());
+        assertThat(allSnapshotsByAll, is(allSnapshots));
+        assertThat(getAllByPatterns(matchAllPattern(), new String[] { snapshotName, weirdSnapshot1, weirdSnapshot2 }), is(allSnapshots));
+        assertThat(getAllByPatterns(new String[] { repoName1, weirdRepo1, weirdRepo2 }, matchAllPattern()), is(allSnapshots));
+
+        assertThat(
+            getAllByPatterns(matchAllPattern(), new String[] { snapshotName }),
+            is(List.of(snapshotInWeird1, snapshotInWeird2, snapshotInRepo1))
+        );
+        assertThat(
+            getAllByPatterns(matchAllPattern(), new String[] { weirdSnapshot1 }),
+            is(List.of(weirdSnapshot1InWeird1, weirdSnapshot1InWeird2, weirdSnapshot1InRepo1))
+        );
+        assertThat(
+            getAllByPatterns(matchAllPattern(), new String[] { snapshotName, weirdSnapshot1 }),
+            is(
+                List.of(
+                    weirdSnapshot1InWeird1,
+                    snapshotInWeird1,
+                    weirdSnapshot1InWeird2,
+                    snapshotInWeird2,
+                    weirdSnapshot1InRepo1,
+                    snapshotInRepo1
+                )
+            )
+        );
+    }
```

Comment:
Can we also verify that `non-existing*,-weird-snapshot-1` does not return `-weird-snapshot-1`? And that '*,--weird-snapshot-1` also does not return it? And that `-*` returns all `-` prefixed snapshots.

Reply:
++ added those cases :)